cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/SourceControlOperationRunner.java

```diff
@@ -33,23 +33,32 @@ public interface SourceControlOperationRunner {
    * @param appToPush     {@link ApplicationDetail} to be pushed
    * @param commitDetails Details of commit author, committer and message
    * @return file-paths and file-hashes for the updated configs.
-   * @throws PushFailureException when the push operation fails for any reason.
+   * @throws NoChangesToPushException      if there is no effective changes on the config file to commit
+   * @throws AuthenticationConfigException if auth config fetch fails
+   * @throws SourceControlFailureException when the push operation fails for any other reason.
    */
   PushAppResponse push(NamespaceId namespace, RepositoryConfig repoConfig,
                        ApplicationDetail appToPush, CommitMeta commitDetails)
-    throws PushFailureException, NoChangesToPushException, AuthenticationConfigException;
+    throws NoChangesToPushException, AuthenticationConfigException;
 
   /**
    * Gets an application spec from a Git repository.
    *
    * @param appRef The {@link ApplicationReference} of the application to pull from
    * @return the details of the pulled application.
-   * @throws PullFailureException          when the operation fails for any reason.
    * @throws NotFoundException             when the requested application is not found in the Git repository.
    * @throws AuthenticationConfigException when there is an error while creating the authentication credentials to
    *                                       call remote Git.
    * @throws IllegalArgumentException      when the fetched application json or file path is invalid.
+   * @throws SourceControlFailureException when the operation fails for any other reason.
    */
-  PullAppResponse<?> pull(ApplicationReference appRef, RepositoryConfig repoConfig) throws PullFailureException,
-    NotFoundException, AuthenticationConfigException;
+  PullAppResponse<?> pull(ApplicationReference appRef, RepositoryConfig repoConfig) throws NotFoundException,
+    AuthenticationConfigException;
+
+  /**
+   * @return Name and git-file-hashes for the detected config files.
+   * @throws SourceControlFailureException when the push operation fails for any reason.
```

Comment:
when would it throw NotFoundException?

Reply:
when the given path-prefix is missing in the repository